iceoryx_hoofs/source/posix_wrapper/periodic_timer.cpp

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2021 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_hoofs/posix_wrapper/periodic_timer.hpp"
+#include "iceoryx_hoofs/posix_wrapper/posix_call.hpp"
+
+using namespace iox::units::duration_literals;
+
+namespace iox
+{
+namespace posix
+{
+PeriodicTimer::PeriodicTimer(const iox::units::Duration interval) noexcept
+    : m_interval(interval)
+{
+    auto timerSemaphore = posix::Semaphore::create(posix::CreateUnnamedSharedMemorySemaphore, 0U);
+    cxx::Ensures(!timerSemaphore.has_error() && "Could not create Semaphore for PeriodicTimer!");
+    m_waitSemaphore = std::move(timerSemaphore.value());
+    start();
+}
+
+void PeriodicTimer::start() noexcept
+{
+    stop();
+    auto waitResult = m_waitSemaphore.timedWait(m_interval);
+    cxx::Ensures(!waitResult.has_error());
+    m_timeForNextActivation = now().value() + m_interval;
+}
+
+void PeriodicTimer::start(const iox::units::Duration interval) noexcept
+{
+    m_interval = interval;
+    start();
+}
+
+void PeriodicTimer::stop() noexcept
+{
+    if (*(m_waitSemaphore.getValue()) == static_cast<int>(posix::SemaphoreWaitState::TIMEOUT))
```

Comment:
I am sorry to say this but this makes no sense. `getValue` returns the current value of the semaphore for instance 5 and casting  `posix::SemaphoreWaitState::TIMEOUT` to an integer provides you with an arbitrary number since there are no values defined for that enum.

Before you continue with this PR please
- read `man sem_overview` (for a technical overview of semaphores)
- read the API of our semaphore
- get familiar with C++ enum classes.

Reply:
Sorry abt mishandling of the enum classes as normal enums. Modified this.